README.md
# GENESIS2PCE

This adapter PCB allows connecting Genesis 3 or 6 button controllers to your PC Engine.
It consists of all the components a real PC Engine 6 button controller has inside of it + an Atmega328p for interfacing with Genesis controllers.

When connecting a Genesis 6 button controller without holding the Mode button, the adapter will go into 6 button mode. This means you can use it as an actual 6 button PC Engine adapter for the handful of games that support it (Street Fighter 2, Art of Fighting, Fatal Fury 2 etc.).
To ensure compatiblity with other games and the Sega Genesis 6 button controller, one should hold the Mode button while powering on the adapter to put the controller into 3 button mode and the adapter into 2 button mode.
Sega Genesis 3 button controllers will always put the adapter into 2 button mode.

- Gerber files can be found in the Gerber directory
- BOMs can be found in the BOM directory
- Hex files for flashing the Atmega328p can be found in the Hex directory

![PCB](https://github.com/Infinest/GENESIS2PCE/blob/master/Images/boardv3.jpg)

![PHOTO](https://github.com/Infinest/GENESIS2PCE/blob/master/Images/boardv3_real.jpg)

# Mapping
| PC Engine | Sega Genesis 3 button | Sega Genesis 6 button|
|--|--|--|
| Up | Up | Up |
| Down| Down | Down |
| Left| Left | Left |
| Right| Right | Right |
| Select | Start / A | Start / Mode|
| Run | A / Start | Mode / Start |
| I | C | C |
| II | B | B |
| III |--| A |
| IV |--| X |
| V |--| Y |
| VI |--| Z |

The button assignments for Select and Run can be swapped. To do this, simply hold A, B, C & Start on your  connected Genesis controller for 8 seconds. This configuration will be retained through power cycles.

# Flashing the Atmega328p

The hex files for flashing the Atmega328p are available inside the Hex directory.
The Atmega328p's fuse bytes should be set to the following values:
| Fuse Byte | Value |
|--|--|
|Fuse Low Byte| 0xDF |
|Fuse High Byte| 0xFE |
|Extended Fuse Byte| 0xFC |

# References
Very helpful information on how to read Genesis controllers was gathered from these sources:
* https://github.com/jonthysell/SegaController
* https://segaretro.org/Six_Button_Control_Pad_(Mega_Drive)#Technical_information 


Hex/Changelog.md
# Changelog
### v1
- Initial code for prototype board
---
### v2
- Change pinout to fit board revision 2
---
### v3
- Change pinout to fit board revision 3
---
### v3.1
- Lower delay inbetween read cycles to 2 ms
----
### v3.2
- Add button combination to swap Run and Select assignment (A, B, C & Start)

millis.c
/*
 * millis.c
 *
 * Created: 02.05.2020 14:42:24
 *  Author: Infinest
 */ 

#include "millis.h"

ISR(TIMER0_OVF_vect)
{
		// copy these to local variables so they can be stored in registers
		// (volatile variables must be read from memory on every access)
		unsigned long m = timer0_millis;
		unsigned int f = timer0_fract;
		m += MILLISECOND_INCREMENT;
		f += FRACT_INCREMENT;
		//Once fraction is big enough to make up an entire millisecond we increment m by one and decrement f by 1000
		if(1000 <= f)
		{
			m++;
			f -= 1000;
		}
		timer0_millis = m;
		timer0_fract = f;
}

void initTimer()
{
	//Enable global interrupts
	sei();
	//Prescale by Clock / 64
	TCCR0B |= 1<<(CS01) | 1<<CS00;
	//Timer/Counter0 Overflow interrupt enable
	TIMSK0  |= 1<<TOIE0;
}

unsigned long millis()
{
	unsigned long m;
	uint8_t oldSREG = SREG;
	cli();
	m = timer0_millis;
	SREG = oldSREG;
	return m;
}

Debug/main.d
main.d main.o: .././main.c .././main.h .././millis.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\interrupt.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\io.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\sfr_defs.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\inttypes.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\lib\gcc\avr\5.4.0\include\stdint.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\stdint.h \
 F:\Programs\Atmel\ Studio\7.0\Packs\atmel\ATmega_DFP\1.3.300\include/avr/iom328p.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\portpins.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\common.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\version.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\fuse.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\lock.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\lib\gcc\avr\5.4.0\include\stdbool.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\util\delay.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\util\delay_basic.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\math.h

.././main.h:

.././millis.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\interrupt.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\io.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\sfr_defs.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\inttypes.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\lib\gcc\avr\5.4.0\include\stdint.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\stdint.h:

F:\Programs\Atmel\ Studio\7.0\Packs\atmel\ATmega_DFP\1.3.300\include/avr/iom328p.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\portpins.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\common.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\version.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\fuse.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\lock.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\lib\gcc\avr\5.4.0\include\stdbool.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\util\delay.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\util\delay_basic.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\math.h:


Release/millis.d
millis.d millis.o: .././millis.c .././millis.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\interrupt.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\io.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\sfr_defs.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\inttypes.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\lib\gcc\avr\5.4.0\include\stdint.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\stdint.h \
 F:\Programs\Atmel\ Studio\7.0\Packs\atmel\ATmega_DFP\1.3.300\include/avr/iom328p.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\portpins.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\common.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\version.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\fuse.h \
 f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\lock.h

.././millis.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\interrupt.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\io.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\sfr_defs.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\inttypes.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\lib\gcc\avr\5.4.0\include\stdint.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\stdint.h:

F:\Programs\Atmel\ Studio\7.0\Packs\atmel\ATmega_DFP\1.3.300\include/avr/iom328p.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\portpins.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\common.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\version.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\fuse.h:

f:\programs\atmel\ studio\7.0\toolchain\avr8\avr8-gnu-toolchain\avr\include\avr\lock.h:


GENESIS2PCE/main.c
/*
* GENESIS2PCE.c
*
* Created: 02.05.2020 14:36:24
* Author : Infinest
*/

#include "main.h"

int main(void)
{
	/*
		Initialize Ports
		PORT C is used for Inputs from Genesis controller and Genesis select pin as output
		PORT B is used for PCE mode output and PCE pad output pins. PB5 is used as select output on Genesis Pin 7
		PORT D is exclusively used for PCE pad output pins
	*/
	DDRC = 0x0;
	DDRB = 0xFF;
	DDRD = 0xFF;
	
	PORTC = 0xFF;
	PORTB = 0xEF; // LED should be off and PCE mode on 2-Button on init
	PORTD = 0xFF;
	
	// Load state of swapping of Run and Select assignment. If value is not 0, make sure it's 1 by doing a bitwise and with 1
	swapState = eeprom_read_byte(0) & 1;
	configureSwappableButtons();

	// Initialize Timer0
	initTimer();
	while (1)
	{
		if(millis() - lastReadTime > MS_DELAY)
		{
			cli();
			readController();
			bitWrite(PORTB, PCE_MODE_AND_LED, sixButtonMode);
			setPCE();
			// If A, B, C and Start are pressed, increment swapTimer. Else reset swapTimer to 0.
			if((state & 0b11110000) == 0)
			{
				swapTimer++;
				// If swapTimer has reached it's trigger value, flip swapState, write it to the eeprom storage and update PC Engine output pin assignment.
				if(swapTimer == SWAP_TIMER_TRIGGER_VALUE)
				{
					swapTimer = 0;
					swapState ^= 1;
					eeprom_write_byte(0, swapState);
					configureSwappableButtons();
				}
			}
			else
			{
				swapTimer = 0;
			}
		}
		sei();
	}
}

/*
	Assigns PC Engine output pins depending on current swapState
*/
void configureSwappableButtons()
{
	if(swapState == 0)
	{
		PC_SWAPPABLE_1 = PC_PIN_SELECT;
		PC_SWAPPABLE_2 = PC_PIN_RUN;
	}
	else
	{
		PC_SWAPPABLE_1 = PC_PIN_RUN;
		PC_SWAPPABLE_2 = PC_PIN_SELECT;

	}
}

/*
	Sets PCE buttons according to current pad mode
*/
void setPCE()
{
	bitWrite(PORTD, PC_PIN_I, bitRead(state, STATE_C));
	bitWrite(PORTD, PC_PIN_II, bitRead(state, STATE_B));
	bitWrite(PORTB, PC_PIN_LEFT, bitRead(state, STATE_LEFT));
	bitWrite(PORTD, PC_PIN_RIGHT, bitRead(state, STATE_RIGHT));
	bitWrite(PORTB, PC_SWAPPABLE_2, bitRead(state, STATE_START));
	bitWrite(PORTD, PC_PIN_UP, bitRead(state, STATE_UP));
	bitWrite(PORTB, PC_PIN_DOWN, bitRead(state, STATE_DOWN));
	if(sixButtonMode)
	{
		bitWrite(PORTB, PC_SWAPPABLE_1, bitRead(state, STATE_MODE));
		bitWrite(PORTD, PC_PIN_III, bitRead(state, STATE_A));
		bitWrite(PORTD, PC_PIN_IV, bitRead(state, STATE_X));
		bitWrite(PORTD, PC_PIN_V, bitRead(state, STATE_Y));
		bitWrite(PORTD, PC_PIN_VI, bitRead(state, STATE_Z));
	}
	else
	{
		bitWrite(PORTB, PC_SWAPPABLE_1, bitRead(state, STATE_A));
		bitWrite(PORTD, PC_PIN_III, 1);
		bitWrite(PORTD, PC_PIN_IV, 1);
		bitWrite(PORTD, PC_PIN_V, 1);
		bitWrite(PORTD, PC_PIN_VI, 1);
	}
}

/*
	Performs one controller read.
	One controller read consists of 8 read cycles.
	See "https://segaretro.org/Six_Button_Control_Pad_(Mega_Drive)#Technical_information"
*/
void readController()
{
	for(uint8_t cycle = 1; cycle <= CYCLES_PER_READ; cycle++)
	{
		if(!doReadCycle(cycle))
		{
			break;
		}
	}
}

/*
	Performs a single read cycle for the Genesis Pad
*/
bool doReadCycle(uint8_t cycle)
{
	bitWrite(PORTB, GEN_7, cycle % 2);
	switch(cycle)
	{
		case 2:
		if((GEN_3 | GEN_4) == 0)
		{
			bitWrite(state, STATE_A, GEN_6);
			bitWrite(state, STATE_START, GEN_9);
		}
		else
		{
			state = 0b1111111111111111;
			sixButtonMode = false;
			return false;
		}
		break;
		case 3:
		bitWrite(state, STATE_UP, GEN_1);
		bitWrite(state, STATE_DOWN, GEN_2);
		bitWrite(state, STATE_LEFT, GEN_3);
		bitWrite(state, STATE_RIGHT, GEN_4);
		bitWrite(state, STATE_B, GEN_6);
		bitWrite(state, STATE_C, GEN_9);
		break;
		case 4:
		_delay_us(1); // MCU is reading the controller too fast. Unless we wait a microsecond we do not get the correct information about the six button mode
		sixButtonMode = GEN_1 == 0 && GEN_2 == 0;
		break;
		case 5:
		if(sixButtonMode)
		{
			bitWrite(state, STATE_Z, GEN_1);
			bitWrite(state, STATE_Y, GEN_2);
			bitWrite(state, STATE_X, GEN_3);
			bitWrite(state, STATE_MODE, GEN_4);
		}
		else
		{
			state |= 0b0000111100000000;
		}
		break;
		case 8:
		lastReadTime = millis();
		break;
	}
	return true;
}


GENESIS2PCE/main.h
/*
 * main.h
 *
 * Created: 02.05.2020 17:10:04
 *  Author: Infinest
 */ 


#ifndef MAIN_H_
#define MAIN_H_

#include "millis.h"
#include <stdint.h>
#include <stdbool.h>
#include <avr/interrupt.h>
#include <avr/io.h>
#include <avr/fuse.h>
#include <util/delay.h>
#include <avr/eeprom.h>

FUSES =
{
	.low = 0xDF,
	.high = 0xFE,
	.extended = 0xFC,
};

#define bitRead(value, bit) (((value) >> (bit)) & 0x01)
#define bitSet(value, bit) ((value) |= (1UL << (bit)))
#define bitClear(value, bit) ((value) &= ~(1UL << (bit)))
#define bitWrite(value, bit, bitvalue) (bitvalue ? bitSet(value, bit) : bitClear(value, bit))

#define MS_DELAY  2
#define CYCLES_PER_READ 8

#define GEN_1 bitRead(PINC, PINC3)
#define GEN_2 bitRead(PINC, PINC2)
#define GEN_3 bitRead(PINC, PINC1)
#define GEN_4 bitRead(PINC, PINC0)
#define GEN_6 bitRead(PINC, PINC5)
#define GEN_7 PINB5
#define GEN_9 bitRead(PINC, PINC4)

#define PCE_MODE_AND_LED PINB4

//1Y
#define PC_PIN_I PIND0//PIND5
#define PC_PIN_UP PIND6//PINB1
//2Y
#define PC_PIN_II PIND1//PIND4
#define PC_PIN_RIGHT PIND7//PINB3
//4Y
#define PC_PIN_RUN PINB3//PIND7
#define PC_PIN_LEFT PINB1//PIND6
//3Y
#define PC_PIN_SELECT PINB2//PINB0
#define PC_PIN_DOWN PINB0//PINB2
// 2 1Y
#define PC_PIN_III PIND2//PIND3
// 2 2Y
#define PC_PIN_IV PIND3//PIND2
// 2 3Y
#define PC_PIN_V PIND4//PIND1
// 2 4Y
#define PC_PIN_VI PIND5//PIND0

#define SWAP_TIMER_TRIGGER_VALUE 0xA00

enum
{
	STATE_UP                      = 0,
	STATE_DOWN                    = 1,
	STATE_LEFT                    = 2,
	STATE_RIGHT                   = 3,
	STATE_START                   = 4,
	STATE_A                       = 5,
	STATE_B                       = 6,
	STATE_C                       = 7,
	STATE_X                       = 8,
	STATE_Y                       = 9,
	STATE_Z                       = 10,
	STATE_MODE                    = 11,
};

unsigned long lastReadTime = 0;
uint16_t state = 0b1111111111111111;
bool sixButtonMode = false;
uint8_t swapState;
uint8_t PC_SWAPPABLE_1;
uint8_t PC_SWAPPABLE_2;
uint16_t swapTimer = 0;

int main(void);
void setPCE();
bool doReadCycle(uint8_t cycle);
void readController();
void writeBit(uint16_t *port, uint8_t pin, uint8_t bit);
void configureSwappableButtons();

#endif /* MAIN_H_ */
